src/wallet/dump.cpp

```diff
@@ -87,3 +87,171 @@ bool DumpWallet(std::shared_ptr<CWallet> wallet, bilingual_str& error)
+        std::string key;
+        std::getline(dump_file, key, ',');
+        std::string value;
+        std::getline(dump_file, value, '\n');
+
+        if (key == "checksum") {
+            std::vector<unsigned char> parsed_checksum = ParseHex(value);
+            std::copy(parsed_checksum.begin(), parsed_checksum.end(), checksum.begin());
+            break;
+        }
+
+        std::string line = strprintf("%s,%s\n", key, value);
+        hasher.write(line.data(), line.size());
+
+        if (key.empty() || value.empty()) {
+            continue;
+        }
+
+        // Check magic for first record
+        if (first) {
+            if (key != DUMP_MAGIC) {
+                error = strprintf(_("Error: Dumpfile identifier record is incorrect. Got \"%s\", expected \"%s\"."), key, DUMP_MAGIC);
+                ret = false;
+                break;
+            }
+            // Check the version number (value of first record)
+            uint32_t ver;
+            if (!ParseUInt32(value, &ver)) {
+                error =strprintf(_("Error: Unable to parse version %u as a uint32_t"), value);
+                ret = false;
+                break;
+            }
+            if (ver != DUMP_VERSION) {
+                error = strprintf(_("Error: Dumpfile version is not supported. This version of bitcoin-wallet only supports version 1 dumpfiles. Got dumpfile with version %s"), value);
+                ret = false;
+                break;
+            }
+            first = false;
+            continue;
+        }
+
+        if (!IsHex(key)) {
+            error = strprintf(_("Error: Got key that was not hex: %s"), key);
+            ret = false;
+            break;
+        }
+        if (!IsHex(value)) {
+            error = strprintf(_("Error: Got value that was not hex: %s"), value);
+            ret = false;
+            break;
+        }
+
+        std::vector<unsigned char> k = ParseHex(key);
+        std::vector<unsigned char> v = ParseHex(value);
+
+        CDataStream ss_key(k, SER_DISK, CLIENT_VERSION);
+        CDataStream ss_value(v, SER_DISK, CLIENT_VERSION);
+
+        if (!batch->Write(ss_key, ss_value)) {
+            error = strprintf(_("Error: Unable to write record to new wallet"));
+            ret = false;
+            break;
+        }
+    }
+
+    if (ret) {
+        uint256 comp_checksum = hasher.GetHash();
+        if (checksum.IsNull()) {
+            error = _("Error: Missing checksum");
+            ret = false;
+        } else if (checksum != comp_checksum) {
+            error = strprintf(_("Error: Dumpfile checksum does not match. Computed %s, expected %s"), HexStr(comp_checksum), HexStr(checksum));
+            ret = false;
+        }
+    }
+
+    if (ret) {
+        batch->TxnCommit();
+    } else {
+        batch->TxnAbort();
+    }
+
+    batch.reset();
+
+    dump_file.close();
+    wallet->Close();
+
+    // Remove the wallet dir if we have a failure
+    if (!ret) {
+        fs::remove_all(wallet_path);
```

Comment:
In commit "wallettool: Add createfromdump command" (fd1f2380cf736df25a27c2607f745ae248c73933)

- I think code probably should call wallet.reset() before this point. `wallet->Close` is called above, but it seems like Close for BDB only flushes the database, so the environment might not get freed until after fs::remove_all, and lead to later attempted writes.
- Alternately this could use a different approach and call a WalletDatabase::Remove method that just removes the known wallet files and empty directory (https://github.com/bitcoin/bitcoin/pull/19137#discussion_r509803825) instead of calling `fs::remove_all`. Even though `remove_all` should be safe because of the `options.require_create = true` line above, a selective remove would seem more safe than a recursive one

Reply:
Added a reset.